pkg/appsec/appsec.go

```diff
@@ -0,0 +1,128 @@
+func newIntakeReverseProxy(target *url.URL, apiKey string, maxPayloadSize int64, transport http.RoundTripper) http.Handler {
+	proxy := httputil.NewSingleHostReverseProxy(target)
+	via := fmt.Sprintf("trace-agent %s", info.Version)
+	// Wrap and overwrite the returned director to add extra headers
+	director := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		// Call the original director changing the request target
+		director(req)
+		// Set extra headers
+		req.Header.Set("Via", via)
+		req.Header.Set("Dd-Api-Key", apiKey)
+		if req.Body != nil && maxPayloadSize > 0 {
+			req.Body = apiutil.NewLimitedReader(req.Body, maxPayloadSize)
+		}
+	}
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusInternalServerError)
+		if err := json.NewEncoder(w).Encode(err); err != nil {
+			log.Error(err)
+		}
+	}
+	proxy.Transport = transport
+	proxy.ErrorLog = stdlog.New(logutil.NewThrottled(5, 10*time.Second), "Appsec backend proxy: ", 0)
+	return withMetrics(proxy)
+}
+
+const (
+	appSecRequestMetricsPrefix            = "datadog.trace_agent.appsec."
+	appSecRequestCountMetricsID           = appSecRequestMetricsPrefix + "request"
+	appSecRequestDurationMetricsID        = appSecRequestMetricsPrefix + "request_duration_ms"
+	appSecRequestPayloadSizeMetricsID     = appSecRequestMetricsPrefix + "request_payload_size"
+	appSecRequestPayloadTooLargeMetricsID = appSecRequestMetricsPrefix + "request_payload_too_large"
+)
+
+func withMetrics(proxy *httputil.ReverseProxy) http.Handler {
+	// Error metrics through the reverse proxy error handler
+	errorHandler := proxy.ErrorHandler
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		if err == apiutil.ErrLimitedReaderLimitReached {
+			if err := metrics.Count(appSecRequestPayloadTooLargeMetricsID, 1, metricsTags(req), 1); err != nil {
+				log.Error(err)
+			}
+		}
+		errorHandler(w, req, err)
+	}
+	// Request metrics through the reverse proxy handler
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		now := time.Now()
+		defer func() {
+			tags := metricsTags(req)
+			if lr, ok := req.Body.(*apiutil.LimitedReader); ok {
+				if err := metrics.Histogram(appSecRequestPayloadSizeMetricsID, float64(lr.Count), tags, 1); err != nil {
+					log.Error(err)
+				}
+			}
+			if err := metrics.Gauge(appSecRequestCountMetricsID, 1, tags, 1); err != nil {
+				log.Error(err)
+			}
+			if err := metrics.Timing(appSecRequestDurationMetricsID, time.Since(now), tags, 1); err != nil {
+				log.Error(err)
+			}
+		}()
+		proxy.ServeHTTP(w, req)
+	})
+}
+
+// metricsTags returns the metrics tags of a request.
+func metricsTags(req *http.Request) []string {
+	tags := []string{"path:" + req.URL.Path}
+	if ct := req.Header.Get("Content-Type"); ct != "" {
+		tags = append(tags, "content_type:"+ct)
+	}
+	if lr, ok := req.Body.(*apiutil.LimitedReader); ok {
```

Comment:
The cardinality for this tag is too high. You'll have to remove it. It's already counted as a metric anyway.

Reply:
Fixed